// pix-utils/dist/module/create.js
import { generatePixObject } from './assembler';
import { PixElementType, } from './types/pixElements';
import { generateErrorObject } from './utils/generateErrorObject';
const defaultPixFields = {
    merchantCategoryCode: '0000',
    transactionCurrency: '986',
    countryCode: 'BR',
};
const defaultStaticFields = {
    ...defaultPixFields,
    isTransactionUnique: false,
    txid: '***',
};
export function createStaticPix(params) {
    if (params.merchantName.length > 25)
        return generateErrorObject('merchantName character limit exceeded (> 25)');
    if (params.txid && params.txid.length > 25)
        return generateErrorObject('txid character limit exceeded (> 25)');
    if (params.merchantCity === '')
        return generateErrorObject('merchantCity is required');
    if (params.merchantCity.length > 15)
        return generateErrorObject('merchantCity character limit exceeded (> 15)');
    const elements = {
        type: PixElementType.STATIC,
        ...defaultStaticFields,
        ...params,
    };
    return generatePixObject(elements);
}
export function createDynamicPix(params) {
    if (params.merchantName.length > 25)
        return generateErrorObject('merchantName character limit exceeded (> 25)');
    if (params.merchantCity === '')
        return generateErrorObject('merchantCity is required');
    if (params.merchantCity.length > 15)
        return generateErrorObject('merchantCity character limit exceeded (> 15)');
    const elements = {
        type: PixElementType.DYNAMIC,
        ...defaultStaticFields,
        ...params,
    };
    return generatePixObject(elements);
}
export function createRecurrencePix(params) {
    if (params.merchantName.length > 25)
        return generateErrorObject('merchantName character limit exceeded (> 25)');
    if (params.merchantCity === '')
        return generateErrorObject('merchantCity is required');
    if (params.merchantCity.length > 15)
        return generateErrorObject('merchantCity character limit exceeded (> 15)');
    const elements = {
        type: PixElementType.RECURRENCE,
        ...defaultStaticFields,
        ...params,
        url: undefined,
    };
    return generatePixObject(elements);
}
