// pix-utils/dist/module/dynamicPayload.js
import axios from 'axios';
import { Buffer } from 'buffer';
import { generateErrorObject } from './utils/generateErrorObject';
import zeroPad from './utils/zeroPad';
export async function fetchPayload({ url, DPP = new Date().toISOString().substring(0, 10), codMun = 5300108, }) {
    const axiosOptions = {
        params: {
            DPP,
            codMun: zeroPad(codMun, 7),
        },
    };
    return axios
        .get(`https://${url}`, axiosOptions)
        .then(({ data, status }) => {
        if (status !== 200)
            return generateErrorObject('Status != 200');
        return data;
    })
        .then((jws) => {
        const parts = jws.split('.').map((b64) => Buffer.from(b64, 'base64'));
        const pixFetch = {
            jwsString: jws,
            jws: {
                hdr: parts[0],
                payload: parts[1],
                signature: parts[2],
            },
            header: JSON.parse(parts[0].toString()),
            payload: JSON.parse(parts[1].toString()),
        };
        return pixFetch;
    })
        .catch((error) => {
        return generateErrorObject(error.message);
    });
}
export async function fetchRecPayload({ url, }) {
    return axios
        .get(`https://${url}`)
        .then(({ data, status }) => {
        if (status !== 200)
            return generateErrorObject('Status != 200');
        return data;
    })
        .then((jws) => {
        const parts = jws.split('.').map((b64) => Buffer.from(b64, 'base64'));
        const pixFetch = {
            jwsString: jws,
            jws: {
                hdr: parts[0],
                payload: parts[1],
                signature: parts[2],
            },
            header: JSON.parse(parts[0].toString()),
            payload: JSON.parse(parts[1].toString()),
        };
        return pixFetch;
    })
        .catch((error) => {
        return generateErrorObject(error.message);
    });
}
