// pix-utils/dist/module/validate.js
import { PixElementType, } from './types/pixElements';
import { EmvMaiSchema, EmvSchema } from './types/pixEmvSchema';
export function isPix(emvElements, test) {
    if (!emvElements.getTag(EmvSchema.TAG_MAI))
        return false;
    const isDynamic = emvElements.getSubTag(EmvMaiSchema.TAG_MAI_URL, EmvSchema.TAG_MAI);
    const isStatic = emvElements.getSubTag(EmvMaiSchema.TAG_MAI_PIXKEY, EmvSchema.TAG_MAI);
    const isRecurrence = emvElements.getSubTag(EmvMaiSchema.TAG_MAI_URL, EmvSchema.TAG_UNRESERVED_TEMPLATE);
    switch (test) {
        case 'pix':
            return true;
        case 'valid':
            return !!isStatic || !!isDynamic || !!isRecurrence;
        case 'static':
            return !!isStatic;
        case 'dynamic':
            return !!isDynamic;
        case 'recurrence':
            return !!isRecurrence;
        default:
            return false;
    }
}
export function hasError(pixElement) {
    return !!pixElement.error;
}
export function hasElementError(pixElement) {
    return !!pixElement.error;
}
export function isStaticPix(pixElement) {
    return pixElement && pixElement.type === PixElementType.STATIC;
}
export function isDynamicPix(pixElement) {
    return pixElement && pixElement.type === PixElementType.DYNAMIC;
}
export function isRecurrencePix(pixElement) {
    return pixElement && pixElement.type === PixElementType.RECURRENCE;
}
