// pix-utils/dist/main/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createStaticPix = createStaticPix;
exports.createDynamicPix = createDynamicPix;
exports.createRecurrencePix = createRecurrencePix;
const assembler_1 = require("./assembler");
const pixElements_1 = require("./types/pixElements");
const generateErrorObject_1 = require("./utils/generateErrorObject");
const defaultPixFields = {
    merchantCategoryCode: '0000',
    transactionCurrency: '986',
    countryCode: 'BR',
};
const defaultStaticFields = Object.assign(Object.assign({}, defaultPixFields), { isTransactionUnique: false, txid: '***' });
function createStaticPix(params) {
    if (params.merchantName.length > 25)
        return (0, generateErrorObject_1.generateErrorObject)('merchantName character limit exceeded (> 25)');
    if (params.txid && params.txid.length > 25)
        return (0, generateErrorObject_1.generateErrorObject)('txid character limit exceeded (> 25)');
    if (params.merchantCity === '')
        return (0, generateErrorObject_1.generateErrorObject)('merchantCity is required');
    if (params.merchantCity.length > 15)
        return (0, generateErrorObject_1.generateErrorObject)('merchantCity character limit exceeded (> 15)');
    const elements = Object.assign(Object.assign({ type: pixElements_1.PixElementType.STATIC }, defaultStaticFields), params);
    return (0, assembler_1.generatePixObject)(elements);
}
function createDynamicPix(params) {
    if (params.merchantName.length > 25)
        return (0, generateErrorObject_1.generateErrorObject)('merchantName character limit exceeded (> 25)');
    if (params.merchantCity === '')
        return (0, generateErrorObject_1.generateErrorObject)('merchantCity is required');
    if (params.merchantCity.length > 15)
        return (0, generateErrorObject_1.generateErrorObject)('merchantCity character limit exceeded (> 15)');
    const elements = Object.assign(Object.assign({ type: pixElements_1.PixElementType.DYNAMIC }, defaultStaticFields), params);
    return (0, assembler_1.generatePixObject)(elements);
}
function createRecurrencePix(params) {
    if (params.merchantName.length > 25)
        return (0, generateErrorObject_1.generateErrorObject)('merchantName character limit exceeded (> 25)');
    if (params.merchantCity === '')
        return (0, generateErrorObject_1.generateErrorObject)('merchantCity is required');
    if (params.merchantCity.length > 15)
        return (0, generateErrorObject_1.generateErrorObject)('merchantCity character limit exceeded (> 15)');
    const elements = Object.assign(Object.assign(Object.assign({ type: pixElements_1.PixElementType.RECURRENCE }, defaultStaticFields), params), { url: undefined });
    return (0, assembler_1.generatePixObject)(elements);
}
