// pix-utils/dist/main/dynamicPayload.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchPayload = fetchPayload;
exports.fetchRecPayload = fetchRecPayload;
const axios_1 = __importDefault(require("axios"));
const buffer_1 = require("buffer");
const generateErrorObject_1 = require("./utils/generateErrorObject");
const zeroPad_1 = __importDefault(require("./utils/zeroPad"));
async function fetchPayload({ url, DPP = new Date().toISOString().substring(0, 10), codMun = 5300108, }) {
    const axiosOptions = {
        params: {
            DPP,
            codMun: (0, zeroPad_1.default)(codMun, 7),
        },
    };
    return axios_1.default
        .get(`https://${url}`, axiosOptions)
        .then(({ data, status }) => {
        if (status !== 200)
            return (0, generateErrorObject_1.generateErrorObject)('Status != 200');
        return data;
    })
        .then((jws) => {
        const parts = jws.split('.').map((b64) => buffer_1.Buffer.from(b64, 'base64'));
        const pixFetch = {
            jwsString: jws,
            jws: {
                hdr: parts[0],
                payload: parts[1],
                signature: parts[2],
            },
            header: JSON.parse(parts[0].toString()),
            payload: JSON.parse(parts[1].toString()),
        };
        return pixFetch;
    })
        .catch((error) => {
        return (0, generateErrorObject_1.generateErrorObject)(error.message);
    });
}
async function fetchRecPayload({ url, }) {
    return axios_1.default
        .get(`https://${url}`)
        .then(({ data, status }) => {
        if (status !== 200)
            return (0, generateErrorObject_1.generateErrorObject)('Status != 200');
        return data;
    })
        .then((jws) => {
        const parts = jws.split('.').map((b64) => buffer_1.Buffer.from(b64, 'base64'));
        const pixFetch = {
            jwsString: jws,
            jws: {
                hdr: parts[0],
                payload: parts[1],
                signature: parts[2],
            },
            header: JSON.parse(parts[0].toString()),
            payload: JSON.parse(parts[1].toString()),
        };
        return pixFetch;
    })
        .catch((error) => {
        return (0, generateErrorObject_1.generateErrorObject)(error.message);
    });
}
