// pix-utils/dist/main/emvHandler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createEmv = createEmv;
exports.parseEmv = parseEmv;
const crc_1 = require("./crc");
const pixElements_1 = require("./types/pixElements");
const pixEmvSchema_1 = require("./types/pixEmvSchema");
const textParser_1 = require("./utils/textParser");
const zeroPad_1 = __importDefault(require("./utils/zeroPad"));
function generateEmvElement(elementId, value) {
    if (!value)
        return '';
    const parsedElementId = (0, zeroPad_1.default)(elementId, 2);
    const parsedLength = (0, zeroPad_1.default)(value.length, 2);
    return `${parsedElementId}${parsedLength}${value}`;
}
function generateMAI(elements) {
    if (elements.type === pixElements_1.PixElementType.STATIC) {
        return [
            generateEmvElement(pixEmvSchema_1.EmvMaiSchema.TAG_MAI_GUI, pixEmvSchema_1.EmvMaiSchema.BC_GUI),
            generateEmvElement(pixEmvSchema_1.EmvMaiSchema.TAG_MAI_PIXKEY, elements.pixKey),
            generateEmvElement(pixEmvSchema_1.EmvMaiSchema.TAG_MAI_INFO_ADD, elements.infoAdicional),
            generateEmvElement(pixEmvSchema_1.EmvMaiSchema.TAG_MAI_FSS, elements.fss),
        ].join('');
    }
    else if (elements.type === pixElements_1.PixElementType.DYNAMIC ||
        elements.type === pixElements_1.PixElementType.RECURRENCE) {
        return [
            generateEmvElement(pixEmvSchema_1.EmvMaiSchema.TAG_MAI_GUI, pixEmvSchema_1.EmvMaiSchema.BC_GUI),
            generateEmvElement(pixEmvSchema_1.EmvMaiSchema.TAG_MAI_URL, elements.url),
        ].join('');
    }
    return undefined;
}
function generateAdditionalData(txid) {
    return generateEmvElement(pixEmvSchema_1.EmvAdditionalDataSchema.TAG_TXID, txid || '***');
}
function createEmv(elements) {
    if (![
        pixElements_1.PixElementType.STATIC,
        pixElements_1.PixElementType.DYNAMIC,
        pixElements_1.PixElementType.RECURRENCE,
    ].includes(elements.type))
        return 'INVALID';
    const emvElements = [
        generateEmvElement(pixEmvSchema_1.EmvSchema.TAG_INIT, '01'),
        generateEmvElement(pixEmvSchema_1.EmvSchema.TAG_ONETIME, elements.oneTime ? '12' : ''),
        generateEmvElement(pixEmvSchema_1.EmvSchema.TAG_MAI, generateMAI(elements)),
        generateEmvElement(pixEmvSchema_1.EmvSchema.TAG_MCC, elements.merchantCategoryCode),
        generateEmvElement(pixEmvSchema_1.EmvSchema.TAG_TRANSACTION_CURRENCY, elements.transactionCurrency),
        elements.type === pixElements_1.PixElementType.STATIC
            ? generateEmvElement(pixEmvSchema_1.EmvSchema.TAG_TRANSACTION_AMOUNT, elements.transactionAmount > 0
                ? elements.transactionAmount.toFixed(2)
                : '')
            : '',
        generateEmvElement(pixEmvSchema_1.EmvSchema.TAG_COUNTRY_CODE, elements.countryCode),
        generateEmvElement(pixEmvSchema_1.EmvSchema.TAG_MERCHANT_NAME, (0, textParser_1.normalizeName)(elements.merchantName)),
        generateEmvElement(pixEmvSchema_1.EmvSchema.TAG_MERCHANT_CITY, (0, textParser_1.normalizeCity)(elements.merchantCity)),
        generateEmvElement(pixEmvSchema_1.EmvSchema.TAG_ADDITIONAL_DATA, generateAdditionalData(elements.type === pixElements_1.PixElementType.STATIC ? elements.txid : '')),
        elements.urlRec
            ? generateEmvElement(pixEmvSchema_1.EmvSchema.TAG_UNRESERVED_TEMPLATE, [
                generateEmvElement(pixEmvSchema_1.EmvMaiSchema.TAG_MAI_GUI, pixEmvSchema_1.EmvMaiSchema.BC_GUI),
                generateEmvElement(pixEmvSchema_1.EmvMaiSchema.TAG_MAI_URL, elements.urlRec),
            ].join(''))
            : '',
        generateEmvElement(pixEmvSchema_1.EmvSchema.TAG_CRC, '0000'),
    ];
    const generatedEmv = emvElements.join('');
    return generatedEmv.replace(/\w{4}$/, (0, crc_1.computeCRC)(generatedEmv));
}
function parseEmv({ emvCode, currentIndex = 0, currentData = {}, }) {
    const tag = +emvCode.substring(currentIndex, currentIndex + 2);
    const length = Number(emvCode.substring(currentIndex + 2, currentIndex + 4));
    const value = emvCode.substring(currentIndex + 4, currentIndex + 4 + length);
    if (!length || !value.length || length !== value.length) {
        return {
            isValid: false,
            rawTags: currentData,
        };
    }
    const newData = Object.assign(Object.assign({}, currentData), { [tag]: Object.assign({ tag: tag, length: length, value: value }, (Object.values(pixEmvSchema_1.TagsWithSubTags).includes(tag)
            ? { subTags: parseEmv({ emvCode: value }) }
            : {})) });
    if (currentIndex + 4 + length === emvCode.length) {
        return {
            isValid: true,
            rawTags: newData,
            getTag: (tag) => { var _a; return (_a = newData === null || newData === void 0 ? void 0 : newData[Number(tag)]) === null || _a === void 0 ? void 0 : _a.value; },
            getSubTag: (tag, mainTag) => { var _a, _b; return (_b = (_a = newData === null || newData === void 0 ? void 0 : newData[Number(mainTag)]) === null || _a === void 0 ? void 0 : _a.subTags) === null || _b === void 0 ? void 0 : _b.getTag(Number(tag)); },
        };
    }
    else {
        return parseEmv({
            emvCode,
            currentIndex: currentIndex + 4 + length,
            currentData: newData,
        });
    }
}
