// pix-utils/dist/main/parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parsePix = parsePix;
exports.extractMandatoryElements = extractMandatoryElements;
exports.extractElements = extractElements;
const assembler_1 = require("./assembler");
const crc_1 = require("./crc");
const emvHandler_1 = require("./emvHandler");
const pixElements_1 = require("./types/pixElements");
const pixEmvSchema_1 = require("./types/pixEmvSchema");
const generateErrorObject_1 = require("./utils/generateErrorObject");
const validate_1 = require("./validate");
function parsePix(brCode) {
    // Parse EMV Code
    const emvElements = (0, emvHandler_1.parseEmv)({ emvCode: brCode });
    if (!emvElements.isValid)
        return (0, generateErrorObject_1.generateErrorObject)('invalid emv code');
    // Validate CRC16
    const crc = (0, crc_1.computeCRC)(brCode);
    if (crc !== emvElements.getTag(pixEmvSchema_1.EmvSchema.TAG_CRC))
        return (0, generateErrorObject_1.generateErrorObject)('invalid crc');
    // Extract Elements
    const elements = extractElements(emvElements);
    if ((0, validate_1.hasElementError)(elements))
        return (0, generateErrorObject_1.generateErrorObject)(elements.message);
    return (0, assembler_1.generatePixObject)(elements);
}
function extractMandatoryElements(emvElements) {
    return {
        merchantCategoryCode: emvElements.getTag(pixEmvSchema_1.EmvSchema.TAG_MCC),
        transactionCurrency: emvElements.getTag(pixEmvSchema_1.EmvSchema.TAG_TRANSACTION_CURRENCY),
        countryCode: emvElements.getTag(pixEmvSchema_1.EmvSchema.TAG_COUNTRY_CODE),
        merchantName: emvElements.getTag(pixEmvSchema_1.EmvSchema.TAG_MERCHANT_NAME),
        merchantCity: emvElements.getTag(pixEmvSchema_1.EmvSchema.TAG_MERCHANT_CITY),
    };
}
function extractElements(emvElements) {
    const basicElements = extractMandatoryElements(emvElements);
    const isRecurrence = (0, validate_1.isPix)(emvElements, 'recurrence');
    if ((0, validate_1.isPix)(emvElements, 'static')) {
        const amountNumber = +emvElements.getTag(pixEmvSchema_1.EmvSchema.TAG_TRANSACTION_AMOUNT);
        const transactionAmount = !isNaN(amountNumber) ? amountNumber : 0;
        return Object.assign(Object.assign({ type: pixElements_1.PixElementType.STATIC }, basicElements), { pixKey: emvElements.getSubTag(pixEmvSchema_1.EmvMaiSchema.TAG_MAI_PIXKEY, pixEmvSchema_1.EmvSchema.TAG_MAI), transactionAmount, infoAdicional: emvElements.getSubTag(pixEmvSchema_1.EmvMaiSchema.TAG_MAI_INFO_ADD, pixEmvSchema_1.EmvSchema.TAG_MAI), txid: emvElements.getSubTag(pixEmvSchema_1.EmvAdditionalDataSchema.TAG_TXID, pixEmvSchema_1.EmvSchema.TAG_ADDITIONAL_DATA), fss: emvElements.getSubTag(pixEmvSchema_1.EmvMaiSchema.TAG_MAI_FSS, pixEmvSchema_1.EmvSchema.TAG_MAI), urlRec: isRecurrence
                ? emvElements.getSubTag(pixEmvSchema_1.EmvMaiSchema.TAG_MAI_URL, pixEmvSchema_1.EmvSchema.TAG_UNRESERVED_TEMPLATE)
                : undefined });
    }
    if ((0, validate_1.isPix)(emvElements, 'dynamic')) {
        return Object.assign(Object.assign({ type: pixElements_1.PixElementType.DYNAMIC }, basicElements), { url: emvElements.getSubTag(pixEmvSchema_1.EmvMaiSchema.TAG_MAI_URL, pixEmvSchema_1.EmvSchema.TAG_MAI), urlRec: isRecurrence
                ? emvElements.getSubTag(pixEmvSchema_1.EmvMaiSchema.TAG_MAI_URL, pixEmvSchema_1.EmvSchema.TAG_UNRESERVED_TEMPLATE)
                : undefined });
    }
    if (isRecurrence) {
        return Object.assign(Object.assign({ type: pixElements_1.PixElementType.RECURRENCE }, basicElements), { url: undefined, urlRec: emvElements.getSubTag(pixEmvSchema_1.EmvMaiSchema.TAG_MAI_URL, pixEmvSchema_1.EmvSchema.TAG_UNRESERVED_TEMPLATE) });
    }
    if (!(0, validate_1.isPix)(emvElements, 'pix') || !(0, validate_1.isPix)(emvElements, 'valid'))
        return (0, generateErrorObject_1.generateErrorObject)('invalid pix');
    return (0, generateErrorObject_1.generateErrorObject)('error');
}
