// pix-utils/dist/main/validate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isPix = isPix;
exports.hasError = hasError;
exports.hasElementError = hasElementError;
exports.isStaticPix = isStaticPix;
exports.isDynamicPix = isDynamicPix;
exports.isRecurrencePix = isRecurrencePix;
const pixElements_1 = require("./types/pixElements");
const pixEmvSchema_1 = require("./types/pixEmvSchema");
function isPix(emvElements, test) {
    if (!emvElements.getTag(pixEmvSchema_1.EmvSchema.TAG_MAI))
        return false;
    const isDynamic = emvElements.getSubTag(pixEmvSchema_1.EmvMaiSchema.TAG_MAI_URL, pixEmvSchema_1.EmvSchema.TAG_MAI);
    const isStatic = emvElements.getSubTag(pixEmvSchema_1.EmvMaiSchema.TAG_MAI_PIXKEY, pixEmvSchema_1.EmvSchema.TAG_MAI);
    const isRecurrence = emvElements.getSubTag(pixEmvSchema_1.EmvMaiSchema.TAG_MAI_URL, pixEmvSchema_1.EmvSchema.TAG_UNRESERVED_TEMPLATE);
    switch (test) {
        case 'pix':
            return true;
        case 'valid':
            return !!isStatic || !!isDynamic || !!isRecurrence;
        case 'static':
            return !!isStatic;
        case 'dynamic':
            return !!isDynamic;
        case 'recurrence':
            return !!isRecurrence;
        default:
            return false;
    }
}
function hasError(pixElement) {
    return !!pixElement.error;
}
function hasElementError(pixElement) {
    return !!pixElement.error;
}
function isStaticPix(pixElement) {
    return pixElement && pixElement.type === pixElements_1.PixElementType.STATIC;
}
function isDynamicPix(pixElement) {
    return pixElement && pixElement.type === pixElements_1.PixElementType.DYNAMIC;
}
function isRecurrencePix(pixElement) {
    return pixElement && pixElement.type === pixElements_1.PixElementType.RECURRENCE;
}
