// pix-utils/dist/module/types/pixDynamicPayload.js
export var PixDynamicStatus;
(function (PixDynamicStatus) {
    PixDynamicStatus["ATIVA"] = "ATIVA";
    PixDynamicStatus["CONCLUIDA"] = "CONCLUIDA";
    PixDynamicStatus["REMOVIDA_PELO_USUARIO_RECEBEDOR"] = "REMOVIDA_PELO_USUARIO_RECEBEDOR";
    PixDynamicStatus["REMOVIDA_PELO_PSP"] = "REMOVIDA_PELO_PSP";
})(PixDynamicStatus || (PixDynamicStatus = {}));
export var Periodicidade;
(function (Periodicidade) {
    Periodicidade["SEMANAL"] = "SEMANAL";
    Periodicidade["MENSAL"] = "MENSAL";
    Periodicidade["TRIMESTRAL"] = "TRIMESTRAL";
    Periodicidade["SEMESTRAL"] = "SEMESTRAL";
    Periodicidade["ANUAL"] = "ANUAL";
})(Periodicidade || (Periodicidade = {}));
export var PoliticaRetentativa;
(function (PoliticaRetentativa) {
    PoliticaRetentativa["NAO_PERMITE"] = "NAO_PERMITE";
    PoliticaRetentativa["PERMITE_3R_7D"] = "PERMITE_3R_7D";
})(PoliticaRetentativa || (PoliticaRetentativa = {}));
export var StatusRec;
(function (StatusRec) {
    StatusRec["CRIADA"] = "CRIADA";
    StatusRec["APROVADA"] = "APROVADA";
    StatusRec["REJEITADA"] = "REJEITADA";
    StatusRec["EXPIRADA"] = "EXPIRADA";
    StatusRec["CANCELADA"] = "CANCELADA";
})(StatusRec || (StatusRec = {}));
export var ModalidadeAgente;
(function (ModalidadeAgente) {
    ModalidadeAgente["AGTEC"] = "AGTEC";
    ModalidadeAgente["AGTOT"] = "AGTOT";
    ModalidadeAgente["AGPSS"] = "AGPSS";
})(ModalidadeAgente || (ModalidadeAgente = {}));
export var ModalidadeAgenteTroco;
(function (ModalidadeAgenteTroco) {
    ModalidadeAgenteTroco["AGTEC"] = "AGTEC";
    ModalidadeAgenteTroco["AGTOT"] = "AGTOT";
})(ModalidadeAgenteTroco || (ModalidadeAgenteTroco = {}));
