// pix-utils/dist/main/types/pixDynamicPayload.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModalidadeAgenteTroco = exports.ModalidadeAgente = exports.StatusRec = exports.PoliticaRetentativa = exports.Periodicidade = exports.PixDynamicStatus = void 0;
var PixDynamicStatus;
(function (PixDynamicStatus) {
    PixDynamicStatus["ATIVA"] = "ATIVA";
    PixDynamicStatus["CONCLUIDA"] = "CONCLUIDA";
    PixDynamicStatus["REMOVIDA_PELO_USUARIO_RECEBEDOR"] = "REMOVIDA_PELO_USUARIO_RECEBEDOR";
    PixDynamicStatus["REMOVIDA_PELO_PSP"] = "REMOVIDA_PELO_PSP";
})(PixDynamicStatus || (exports.PixDynamicStatus = PixDynamicStatus = {}));
var Periodicidade;
(function (Periodicidade) {
    Periodicidade["SEMANAL"] = "SEMANAL";
    Periodicidade["MENSAL"] = "MENSAL";
    Periodicidade["TRIMESTRAL"] = "TRIMESTRAL";
    Periodicidade["SEMESTRAL"] = "SEMESTRAL";
    Periodicidade["ANUAL"] = "ANUAL";
})(Periodicidade || (exports.Periodicidade = Periodicidade = {}));
var PoliticaRetentativa;
(function (PoliticaRetentativa) {
    PoliticaRetentativa["NAO_PERMITE"] = "NAO_PERMITE";
    PoliticaRetentativa["PERMITE_3R_7D"] = "PERMITE_3R_7D";
})(PoliticaRetentativa || (exports.PoliticaRetentativa = PoliticaRetentativa = {}));
var StatusRec;
(function (StatusRec) {
    StatusRec["CRIADA"] = "CRIADA";
    StatusRec["APROVADA"] = "APROVADA";
    StatusRec["REJEITADA"] = "REJEITADA";
    StatusRec["EXPIRADA"] = "EXPIRADA";
    StatusRec["CANCELADA"] = "CANCELADA";
})(StatusRec || (exports.StatusRec = StatusRec = {}));
var ModalidadeAgente;
(function (ModalidadeAgente) {
    ModalidadeAgente["AGTEC"] = "AGTEC";
    ModalidadeAgente["AGTOT"] = "AGTOT";
    ModalidadeAgente["AGPSS"] = "AGPSS";
})(ModalidadeAgente || (exports.ModalidadeAgente = ModalidadeAgente = {}));
var ModalidadeAgenteTroco;
(function (ModalidadeAgenteTroco) {
    ModalidadeAgenteTroco["AGTEC"] = "AGTEC";
    ModalidadeAgenteTroco["AGTOT"] = "AGTOT";
})(ModalidadeAgenteTroco || (exports.ModalidadeAgenteTroco = ModalidadeAgenteTroco = {}));
